src/contracts/atlas/GasAccounting.sol

```diff
@@ -201,107 +208,187 @@ abstract contract GasAccounting is SafetyLocks {
-    /// @dev If borrowing AtlETH is successful, sets the solver lock and returns true; otherwise, returns false.
-    /// @param solverOp The SolverOperation data for the borrowing solver.
-    /// @return valid True if the solver lock is successfully set, otherwise false.
-    function _trySolverLock(SolverOperation calldata solverOp) internal returns (bool valid) {
-        if (_borrow(solverOp.value)) {
-            _solverLock = uint256(uint160(solverOp.from));
-            return true;
-        }
-
-        return false;
-    }
-
     /// @notice Releases the solver lock and adjusts the solver's escrow balance based on the gas used and other
     /// factors.
     /// @dev Calculates the gas used for the SolverOperation and adjusts the solver's escrow balance accordingly.
     /// @param solverOp The current SolverOperation for which to account
     /// @param gasWaterMark The `gasleft()` watermark taken at the start of executing the SolverOperation.
     /// @param result The result bitmap of the SolverOperation execution.
-    /// @param bidFind Indicates if called in the context of `_getBidAmount` in Escrow.sol (true) or not (false).
     /// @param includeCalldata Whether to include calldata cost in the gas calculation.
-    function _releaseSolverLock(
+    function _handleSolverAccounting(
         SolverOperation calldata solverOp,
         uint256 gasWaterMark,
         uint256 result,
-        bool bidFind,
         bool includeCalldata
     )
         internal
     {
-        // Calculate what the solver owes if they failed
-        // NOTE: This will cause an error if you are simulating with a gasPrice of 0
-        if (!bidFind && !result.updateEscrow()) return;
-
-        uint256 gasUsed = (gasWaterMark - gasleft() + _SOLVER_LOCK_GAS_BUFFER) * tx.gasprice;
+        uint256 gasUsed = (gasWaterMark + _SOLVER_BASE_GAS_USED - gasleft()) * tx.gasprice;
 
         if (includeCalldata) {
             gasUsed += _getCalldataCost(solverOp.data.length);
         }
 
-        gasUsed = (gasUsed + ((gasUsed * SURCHARGE_RATE) / SURCHARGE_SCALE));
-        _assign(solverOp.from, gasUsed, false, bidFind);
+        gasUsed = gasUsed * (SURCHARGE_SCALE + ATLAS_SURCHARGE_RATE + BUNDLER_SURCHARGE_RATE) / SURCHARGE_SCALE;
+
+        // Calculate what the solver owes
+        // NOTE: This will cause an error if you are simulating with a gasPrice of 0
+        if (!result.updateEscrow()) {
+            // CASE: Solver is not responsible for the failure of their operation, so we blame the bundler
+            // and reduce the total amount refunded to the bundler
+            writeoffs += gasUsed;
+        } else {
+            // CASE: Solver failed, so we calculate what they owe.
+            uint256 deficit = _assign(solverOp.from, gasUsed, false);
+            if (deficit > 0) {
+                // Write off any deficit as a gas loss to the bundler so that other solvers aren't forced to pay it.
```

Comment:
I thought from discord conversations it wasn't possible for a solver to not have enough atlETH to cover the gasLimit we allocate to their solverCall?

Reply:
It shouldn't be possible. I guess this is an extra check just in case?